components/cdc/src/service.rs

```diff
@@ -195,132 +257,238 @@ impl Service {
+                // Get version from the first request in the stream.
+                let version = Self::parse_version_from_request_header(&request, &peer);
+                Self::set_conn_version(&scheduler, conn_id, version, explicit_features)?;
+                Self::handle_request(&scheduler, &peer, request, conn_id)?;
+            }
+            while let Some(request) = stream.try_next().await? {
+                Self::handle_request(&scheduler, &peer, request, conn_id)?;
+            }
             let deregister = Deregister::Conn(conn_id);
             if let Err(e) = scheduler.schedule(Task::Deregister(deregister)) {
                 error!("cdc deregister failed"; "error" => ?e, "conn_id" => ?conn_id);
             }
-            match res {
-                Ok(()) => {
-                    info!("cdc receive closed"; "downstream" => peer, "conn_id" => ?conn_id);
-                }
-                Err(e) => {
-                    warn!("cdc receive failed"; "error" => ?e, "downstream" => peer, "conn_id" => ?conn_id);
-                }
+            Ok::<(), String>(())
+        };
+
+        let peer = ctx.peer();
+        ctx.spawn(async move {
+            if let Err(e) = recv_req.await {
+                warn!("cdc receive failed"; "error" => ?e, "downstream" => peer, "conn_id" => ?conn_id);
+            } else {
+                info!("cdc receive closed"; "downstream" => peer, "conn_id" => ?conn_id);
             }
         });
 
         let peer = ctx.peer();
-        let scheduler = self.scheduler.clone();
-
         ctx.spawn(async move {
             #[cfg(feature = "failpoints")]
             sleep_before_drain_change_event().await;
-
-            let res = event_drain.forward(&mut sink).await;
-            // Unregister this downstream only.
-            let deregister = Deregister::Conn(conn_id);
-            if let Err(e) = scheduler.schedule(Task::Deregister(deregister)) {
-                error!("cdc deregister failed"; "error" => ?e);
-            }
-            match res {
-                Ok(_s) => {
-                    info!("cdc send closed"; "downstream" => peer, "conn_id" => ?conn_id);
-                    let _ = sink.close().await;
-                }
-                Err(e) => {
-                    warn!("cdc send failed"; "error" => ?e, "downstream" => peer, "conn_id" => ?conn_id);
-                }
+            if let Err(e) = event_drain.forward(&mut sink).await {
+                warn!("cdc send failed"; "error" => ?e, "downstream" => peer, "conn_id" => ?conn_id);
+            } else {
+                info!("cdc send closed"; "downstream" => peer, "conn_id" => ?conn_id);
             }
         });
     }
 }
 
+impl ChangeData for Service {
+    fn event_feed(
+        &mut self,
+        ctx: RpcContext<'_>,
+        stream: RequestStream<ChangeDataRequest>,
+        sink: DuplexSink<ChangeDataEvent>,
+    ) {
+        self.handle_event_feed(ctx, stream, sink, false);
+    }
+
+    fn event_feed_v2(
```

Comment:
As https://github.com/pingcap/kvproto/pull/1130#discussion_r1230785817, I don't think it's necessary to add a duplicated RPC.

Reply:
Comparing with a new RPC, I think it's more dirty to handle it in TiCDC. Why you are trying so hard to avoid a new RPC, event if it will introduce more extra code, comment and tests?

I decide to let it be, because a new RPC doesn't break anything about the old one. If it leads any problems, I will handle them.

Reply:
Okay, if you insist. 🙂